xir/parser.py

```diff
@@ -275,69 +232,74 @@ def application_stmt(self, args):
         }
         return Statement(name, params, wires, **stmt_options)
 
-    def obs_stmt(self, args):
-        """Observable statement. Defined inside an observable definition.
+    @v_args(inline=True)
+    def obs_def(self, name, params, wires, statements):
 
-        Returns:
-            ObservableStmt: object containing statement data
-        """
-        pref = simplify_math(args[0])
-        terms = args[1]
-        return ObservableStmt(pref, terms, use_floats=self.use_floats)
+        if wires is None:
+            max_wire = 0
+            for stmt in statements:
+                for factor in stmt.factors:
+                    int_wires = [w for w in factor.wires if isinstance(w, int)]
+                    if int_wires and max(int_wires) > max_wire:
+                        max_wire = max(int_wires)
 
-    def obs_group(self, args):
-        """Group of observables used to define an observable statement.
+            wires = tuple(range(max_wire + 1))
+        else:
+            # remove duplicate wires while maintaining order
+            wires = tuple(dict.fromkeys(wires))
 
-        Returns:
-            list[tuple]: each observable with corresponding wires as tuples
-        """
-        return [(args[i], args[i + 1]) for i in range(0, len(args) - 1, 2)]
+        params = tuple(params[1]) if params is not None else tuple()
+        self._program.add_observable(name, params, wires, statements)
+
+    def obs_stmt_list(self, stmts):
+        return stmts
+
+    @v_args(inline=True)
+    def obs_stmt(self, coeff, factors):
+        return ObservableStmt(simplify_math(coeff), factors, use_floats=self.use_floats)
```

Comment:
It would be nice to stick with one nomenclature for the pre-factor/coefficient of an observable statement.
```suggestion
    def obs_stmt(self, pref, factors):
        return ObservableStmt(simplify_math(pref), factors, use_floats=self.use_floats)
```

Reply:
changed